### ISSUE TITLE
vegan::ordiR2step() doesn't find best-fit model

### ISSUE THREAD

**Issue by EDiLD** [vegandevs/vegan#223]
Regarding this question [posted on SO](http://stackoverflow.com/questions/41937957/veganordir2step-doesnt-find-best-fit-model):

I totally agree with your answer, however, there is this error mentioned:

```
Error in terms.formula(tmp, simplify = TRUE) : 
  invalid model formula in ExtractVars
```

It originates because in line 87 of [ordiR2step](https://github.com/vegandevs/vegan/blob/master/R/ordiR2step.R)  `scope` resolves to `NA`.

I think this error should no pop up (?).
Maybe there is an additional check needed (if R2scope = `FALSE`), like

`best <= length(adds)`
or 
`substr(names(best), 1, 1) != '-'`

Any thoughts on this?

**Comment by jarioksa** [comment]
I'm currently writing a fix on this.

The problem was that dropping was not handled correctly, but the function always tried to add terms (and it indeed contains that  `substr` command to fork the process).

**Comment by EDiLD** [comment]
Great that you picked this up (wasn't clear to me from your answer on SO)!

**Comment by jarioksa** [comment]
PR #224 may fix the problems. However, the function is messy and the logic is obscure. I really don't like the function like it stands now.

I think the easiest way of making this cleaner is to have this as forward-only. That is, we never consider a variable for removal.

**Comment by EDiLD** [comment]
> I think the easiest way of making this cleaner is to have this as forward-only. That is, we never consider a variable for removal.

:+1:

**Comment by jarioksa** [comment]
With later commits, the PR #224  now has code where `ordiR2step` only works forward, and the `direction` argument was completely dropped. The code is now much simpler (and I also corrected some omissions, like returning the `anova` component).

The logic of stepwise (`forward`/`backward`) code was that we have three alternative stopping criteria. In the current `forward` advance we stop when any of these criteria is met (but two can be turned off by `R2scope` and `Pin` arguments). In `backward`  model building the logic was not as clear: we always consider only the topmost term which also for the removal would be the best, and we stopped when any of the criteria was met for the topmost term. This means that worse terms could remain in the model and never considered. (In theory: in practice the `backward` stage never really worked except in some version of the PR #224 branch.)
